oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java

```diff
@@ -244,73 +255,83 @@ public FlatFileStore build() throws IOException, CompositeException {
                     unmodifiableSet(preferredPathElements), algorithm);
             storeList.add(store);
         }
         return storeList;
     }
 
     /**
-     * Returns the existing list of store files if it can read from system property OAK_INDEXER_SORTED_FILE_PATH which 
-     * defines the existing folder where the flat file store files are present. Will throw an exception if it cannot 
+     * Returns the existing list of store files if it can read from system property OAK_INDEXER_SORTED_FILE_PATH which
+     * defines the existing folder where the flat file store files are present. Will throw an exception if it cannot
      * read or the path in the system property is not a directory.
-     * If the system property OAK_INDEXER_SORTED_FILE_PATH in undefined, or it cannot read relevant files it 
+     * If the system property OAK_INDEXER_SORTED_FILE_PATH in undefined, or it cannot read relevant files it
      * initializes the flat file store.
-     * 
-     * @return list of flat files
-     * @throws IOException 
+     *
+     * @return pair of "list of flat files" and metadata file
+     * @throws IOException
      * @throws CompositeException
      */
-    private List<File> createdSortedStoreFiles() throws IOException, CompositeException {
+    private Pair<List<File>, List<File>> createdSortedStoreFiles() throws IOException, CompositeException {
         // Check system property defined path
         String sortedFilePath = System.getProperty(OAK_INDEXER_SORTED_FILE_PATH);
         if (StringUtils.isNotBlank(sortedFilePath)) {
             File sortedDir = new File(sortedFilePath);
             log.info("Attempting to read from provided sorted files directory [{}] (via system property '{}')",
-                sortedDir.getAbsolutePath(), OAK_INDEXER_SORTED_FILE_PATH);
-            List<File> files = getFiles(sortedDir);
-            if (files != null) {
-                return files;
+                    sortedDir.getAbsolutePath(), OAK_INDEXER_SORTED_FILE_PATH);
+            // List of storefiles, List of metadatafile
+            Pair<List<File>, List<File>> storeFiles = getIndexStoreFiles(sortedDir);
+            if (storeFiles != null) {
+                return storeFiles;
             }
         }
 
         // Initialize the flat file store again
 
         createStoreDir();
-        SortStrategy strategy = createSortStrategy(flatFileStoreDir);
+        IndexStoreSortStrategy strategy = createSortStrategy(flatFileStoreDir);
         File result = strategy.createSortedStoreFile();
+        File metadata = strategy.createMetadataFile();
         entryCount = strategy.getEntryCount();
-        return Collections.singletonList(result);
+        return new ImmutablePair<>(Collections.singletonList(result), Collections.singletonList(metadata));
     }
 
     @Nullable
-    private List<File> getFiles(File sortedDir) {
+    private Pair<List<File>, List<File>> getIndexStoreFiles(File sortedDir) {
```

Comment:
Reading the return type, it is not immediately obvious what the distinction between the two lists is. Can we either use a custom data class like: 
```java
public class IndexStoreFiles {
    private final List<File> storeFiles;
    private final List<File> metadataFiles;

    public IndexStoreFiles(List<File> storeFiles, List<File> metadataFiles) {
        this.storeFiles = storeFiles;
        this.metadataFiles = metadataFiles;
    }

    public List<File> getStoreFiles() {
        return storeFiles;
    }

    public List<File> getMetadataFiles() {
        return metadataFiles;
    }
}
```

or a `Map<String, List<File>>`?

Reply:
I'm also not quite sure... 

If we have one metafile per store file (?) then we could also use List<Pair<File, File>>.

But a separate class might be even better.

Reply:
I wonder if we really need to have a List<File> metadataFiles. Isn't the metadata file always the same as the storedFile, except for an added suffix?

Reply:
created an inner class.